migrate/migrate.go

```diff
@@ -113,10 +113,84 @@ func FindMigrations(path string) ([]string, error) {
 	return paths, nil
 }
 
+
+// MigrationDir is a function that follows the semantics defined by the go-bindata AssetDir function:
+// AssetDir returns the file names below a certain
+// returns the file names below a certain
+// directory embedded in the file by go-bindata.
+// For example if you run go-bindata on data/... and data contains the
+// following hierarchy:
+//     data/
+//       foo.txt
+//       img/
+//         a.png
+//         b.png
+// then AssetDir("data") would return []string{"foo.txt", "img"}
+// AssetDir("data/img") would return []string{"a.png", "b.png"}
+// AssetDir("foo.txt") and AssetDir("notexist") would return an error
+// AssetDir("") will return []string{"data"}.
+type MigrationDir func (string) ([]string, error)
+
+// MigrationAsset is a function that follows the semantics defined by the go-bindata Asset function:
+// Asset loads and returns the asset for the given name.
+// It returns an error if the asset could not be found or
+// could not be loaded.
+type MigrationAsset func (string) ([]byte, error)
+
+// recursive function to harvest the shared sql templates.
+func traverseShared(l migrationLoader, dir MigrationDir, asset MigrationAsset, cwd string, sps []string) error {
+	for _, name := range sps {
+		filePath := filepath.Join(cwd, name)
+		if sharedDirs, err := dir(name); err == nil {
+			traverseShared(l, dir, asset, filePath, sharedDirs)
+		} else if body, err := asset(filePath); err != nil {
+			return err
+		} else if err := l.loadShared(name, body); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type kv struct {
+	key string
+	value []byte
+}
+
+func (m * Migrator) LoadMigrationsFromGoBindata(base string, dir MigrationDir, asset MigrationAsset) error {
+	if as, err := dir(base); err != nil {
+		return err
+	} else {
+		toResolve := []kv{}
+		loader := m.newMigrationLoader(base+"/")
+		for _, name := range as {
+			filePath := filepath.Join(base, name)
+			// first check if the entry is a subdirectory, if it is harvest shared sql
+			if shared, err := dir(filePath); err == nil {
+				if err := traverseShared(loader, dir, asset, filePath, shared); err != nil {
+					return err
+				}
+				// otherwise it must be a migration at the root, in which case defer processing it till all
+				// shared sql has been loaded
+			} else if body, err := asset(filePath); err != nil {
+				return err
+			} else {
+				toResolve = append(toResolve, kv{ key: name, value: body})
+			}
+		}
+		// now resolve the actual migrations
+		for k, v := range toResolve {
+			if err := loader.load(k, v); err != nil { return err }
```

Comment:
I had a map instead of the kv struct earlier. I realised I needed the logic from `FindMigrations` here. 
The code is working in our codebase but I am not really using nested shared sql directories or the template support yet.